File: app/src/main/java/com/ramez/shopp/Fragments/CategoryProductsFragment.java
package com.ramez.shopp.Fragments;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.karumi.dexter.Dexter;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionDeniedResponse;
import com.karumi.dexter.listener.PermissionGrantedResponse;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.single.PermissionListener;
import com.ramez.shopp.Activities.FullScannerActivity;
import com.ramez.shopp.Activities.ProductDetailsActivity;
import com.ramez.shopp.Adapter.MainCategoryAdapter;
import com.ramez.shopp.Adapter.ProductCategoryAdapter;
import com.ramez.shopp.Adapter.SubCategoryAdapter;
import com.ramez.shopp.ApiHandler.DataFeacher;
import com.ramez.shopp.Classes.CategoryModel;
import com.ramez.shopp.Classes.Constants;
import com.ramez.shopp.Classes.MessageEvent;
import com.ramez.shopp.Classes.UtilityApp;
import com.ramez.shopp.Models.AutoCompleteModel;
import com.ramez.shopp.Models.ChildCat;
import com.ramez.shopp.Models.FavouriteResultModel;
import com.ramez.shopp.Models.LocalModel;
import com.ramez.shopp.Models.MemberModel;
import com.ramez.shopp.Models.ProductModel;
import com.ramez.shopp.R;
import com.ramez.shopp.databinding.FragmentCategoryProductsBinding;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;

import static android.app.Activity.RESULT_OK;

public class CategoryProductsFragment extends FragmentBase implements ProductCategoryAdapter.OnItemClick, MainCategoryAdapter.OnMainCategoryItemClicked {
    private static final int ZBAR_CAMERA_PERMISSION = 1;
    ArrayList<ProductModel> productList;
    ArrayList<AutoCompleteModel> data = null;
    ArrayList<String> autoCompleteList;
    ArrayList<CategoryModel> mainCategoryDMS;
    GridLayoutManager gridLayoutManager;
    int numColumn = 2;
    int selectedSubCat = 0;
    int category_id = 0, country_id, city_id;
    private FragmentCategoryProductsBinding binding;
    private String user_id = "0", filter = "";
    private MemberModel user;
    private LocalModel localModel;
    private ProductCategoryAdapter adapter;
    private int SEARCH_CODE = 2000;


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentCategoryProductsBinding.inflate(inflater, container, false);

        productList = new ArrayList<>();
        mainCategoryDMS = new ArrayList<>();


        gridLayoutManager = new GridLayoutManager(getActivityy(), numColumn);
        binding.productsRv.setLayoutManager(gridLayoutManager);


        binding.listShopCategories.setLayoutManager(new LinearLayoutManager(getActivityy(), LinearLayoutManager.HORIZONTAL, false));

        binding.listSubCategory.setLayoutManager(new LinearLayoutManager(getActivityy(), LinearLayoutManager.HORIZONTAL, false));


        binding.productsRv.setHasFixedSize(true);
        binding.productsRv.setItemAnimator(null);

        data = new ArrayList<>();
        autoCompleteList = new ArrayList<>();

        localModel = UtilityApp.getLocalData();
        country_id = localModel.getCountryId();


        if (UtilityApp.isLogin()) {

            user = UtilityApp.getUserData();
            user_id = String.valueOf(user.getId());

        }
        city_id = Integer.parseInt(localModel.getCityId());

        getIntentExtra();

        binding.failGetDataLY.refreshBtn.setOnClickListener(view1 -> {

            getProductList(selectedSubCat, country_id, city_id, user_id, filter, 0, 10);

        });


        binding.searchBut.setOnClickListener(view1 -> {
            EventBus.getDefault().post(new MessageEvent(MessageEvent.TYPE_search));
            FragmentManager fragmentManager = getParentFragmentManager();
            SearchFragment searchFragment = new SearchFragment();
            fragmentManager.beginTransaction().replace(R.id.mainContainer, searchFragment, "searchFragment").commit();


        });


        binding.priceBut.setOnClickListener(view1 -> {
            Collections.sort(productList, Collections.reverseOrder());

        });


        binding.barcodeBut.setOnClickListener(view1 -> {

            checkCameraPermission();

        });


        binding.categoriesCountTv.setText(String.valueOf(productList.size()));

        binding.offerCountTv.setText(String.valueOf(productList.size()));

        return binding.getRoot();

    }

    public void initAdapter() {
        adapter = new ProductCategoryAdapter(getActivityy(), binding.productsRv, productList, selectedSubCat, country_id, city_id, user_id, 0, "", this, numColumn,0);
        binding.productsRv.setAdapter(adapter);

        binding.categoriesCountTv.setText(String.valueOf(productList.size()));
        binding.offerCountTv.setText(String.valueOf(productList.size()));
    }

    @Override
    public void onItemClicked(int position, ProductModel productModel) {
        Intent intent = new Intent(getActivityy(), ProductDetailsActivity.class);
        intent.putExtra(Constants.DB_productModel, productModel);
        startActivity(intent);


    }

    private void cancelAPiCall() {
        if (adapter != null && adapter.apiCall != null && adapter.apiCall.isExecuted()) {
            adapter.isCanceled = true;
            adapter.apiCall.cancel();
        }
    }

    private void getIntentExtra() {
        Bundle bundle = getArguments();
        Log.i(CategoryProductsFragment.class.getName(), "Log getIntentExtra ");

        if (bundle != null) {

            mainCategoryDMS = (ArrayList<CategoryModel>) bundle.getSerializable(Constants.CAT_LIST);
            category_id = bundle.getInt(Constants.SELECTED_POSITION, 0);
            CategoryModel categoryModel = (CategoryModel) bundle.getSerializable(Constants.CAT_MODEL);
            int position = bundle.getInt(Constants.position, 0);
            Log.i(CategoryProductsFragment.class.getName(), "Log mainCategoryDMS " + mainCategoryDMS.size());
            Log.i(CategoryProductsFragment.class.getName(), "Log category_id " + category_id);
            Log.i(CategoryProductsFragment.class.getName(), "Log categoryModel " + categoryModel.getHName());

            selectedSubCat = categoryModel.getId();
            initMainCategoryAdapter();

            initSubCatList(mainCategoryDMS.get(position).getChildCat());

            getProductList(selectedSubCat, country_id, city_id, user_id, filter, 0, 10);

        }
    }

    private void initSubCatList(ArrayList<ChildCat> subCatList) {
        ChildCat childCat = new ChildCat();
        childCat.setId(selectedSubCat);
        childCat.setHName(getString(R.string.all));
        childCat.setName(getString(R.string.all));

     //   selectedSubCat = category_id;
        ArrayList<ChildCat> subCategoryDMS = new ArrayList<>(subCatList);

        subCategoryDMS.add(0, childCat);

        initSubCategoryAdapter(subCategoryDMS);
    }

    private void initSubCategoryAdapter(ArrayList<ChildCat> subCategoryDMS) {

        SubCategoryAdapter subCategoryAdapter = new SubCategoryAdapter(getActivityy(), subCategoryDMS, object -> {

            selectedSubCat = object.getId();
            cancelAPiCall();
            getProductList(selectedSubCat, country_id, city_id, user_id, "", 0, 10);

        }, selectedSubCat);

        binding.listSubCategory.setAdapter(subCategoryAdapter);

    }

    private void initMainCategoryAdapter() {
        MainCategoryAdapter mainCategoryShopAdapter = new MainCategoryAdapter(getActivityy(), mainCategoryDMS, this, selectedSubCat);
        binding.listShopCategories.setAdapter(mainCategoryShopAdapter);
    }


    public void getProductList(int category_id, int country_id, int city_id, String user_id, String filter, int page_number, int page_size) {
//        productList.clear();
        binding.loadingProgressLY.loadingProgressLY.setVisibility(View.VISIBLE);
        binding.productsRv.setVisibility(View.GONE);
        binding.noDataLY.noDataLY.setVisibility(View.GONE);
        binding.failGetDataLY.failGetDataLY.setVisibility(View.GONE);

        new DataFeacher(false, (obj, func, IsSuccess) -> {
            if (isVisible()) {

                FavouriteResultModel result = (FavouriteResultModel) obj;
                String message = getString(R.string.fail_to_get_data);

                binding.loadingProgressLY.loadingProgressLY.setVisibility(View.GONE);

                if (func.equals(Constants.ERROR)) {

                    if (result != null && result.getMessage() != null) {
                        message = result.getMessage();
                    }
                    binding.productsRv.setVisibility(View.GONE);
                    binding.noDataLY.noDataLY.setVisibility(View.GONE);
                    binding.failGetDataLY.failGetDataLY.setVisibility(View.VISIBLE);
                    binding.failGetDataLY.failTxt.setText(message);

                } else if (func.equals(Constants.NO_CONNECTION)) {
                    binding.failGetDataLY.failGetDataLY.setVisibility(View.VISIBLE);
                    binding.failGetDataLY.failTxt.setText(R.string.no_internet_connection);
                    binding.failGetDataLY.noInternetIv.setVisibility(View.VISIBLE);
                    binding.productsRv.setVisibility(View.GONE);

                } else {
                    if (IsSuccess) {
                        if (result.getData() != null && result.getData().size() > 0) {

                            binding.productsRv.setVisibility(View.VISIBLE);
                            binding.noDataLY.noDataLY.setVisibility(View.GONE);
                            binding.failGetDataLY.failGetDataLY.setVisibility(View.GONE);
                            productList = new ArrayList<>(result.getData());
                            initAdapter();

                        } else {

                            binding.productsRv.setVisibility(View.GONE);
                            binding.noDataLY.noDataLY.setVisibility(View.VISIBLE);

                        }


                    } else {

                        binding.productsRv.setVisibility(View.GONE);
                        binding.noDataLY.noDataLY.setVisibility(View.GONE);
                        binding.failGetDataLY.failGetDataLY.setVisibility(View.VISIBLE);
                        binding.failGetDataLY.failTxt.setText(message);


                    }
                }

            }

        }).getFavorite(category_id, country_id, city_id, user_id, filter, 0, page_number, page_size);

    }

    @Override
    public void OnMainCategoryItemClicked(CategoryModel mainCategoryDM, int position) {

        selectedSubCat = mainCategoryDM.getId();
        initSubCatList(mainCategoryDM.getChildCat());

        cancelAPiCall();

        getProductList(selectedSubCat, country_id, city_id, user_id, filter, 0, 10);


    }


    private void checkCameraPermission() {
        Dexter.withContext(getActivityy()).withPermission(Manifest.permission.CAMERA).withListener(new PermissionListener() {
            @RequiresApi(api = Build.VERSION_CODES.M)
            @Override
            public void onPermissionGranted(PermissionGrantedResponse response) {
                startScan();


            }

            @Override
            public void onPermissionDenied(PermissionDeniedResponse response) {
                Toast.makeText(getActivityy(), "" + getString(R.string.permission_camera_rationale), Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onPermissionRationaleShouldBeShown(PermissionRequest permission, PermissionToken token) {
                token.continuePermissionRequest();

            }
        }).withErrorListener(error -> Toast.makeText(getActivityy(), "" + getString(R.string.error_in_data), Toast.LENGTH_SHORT).show()).onSameThread().check();
    }


    private void startScan() {

        if (ContextCompat.checkSelfPermission(getActivityy(), Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(getActivityy(), new String[]{Manifest.permission.CAMERA}, ZBAR_CAMERA_PERMISSION);
        } else {

            Intent intent = new Intent(getActivityy(), FullScannerActivity.class);
            startActivityForResult(intent, SEARCH_CODE);
        }

    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(@NotNull MessageEvent event) {
        if (event.type.equals(MessageEvent.TYPE_view)) {
            numColumn = (int) event.data;
            initAdapter();
            gridLayoutManager.setSpanCount(numColumn);
            adapter.notifyDataSetChanged();


        }
    }


    @Override
    public void onStart() {
        super.onStart();
        EventBus.getDefault().register(this);

    }


    @Override
    public void onStop() {
        super.onStop();
        EventBus.getDefault().unregister(this);

    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK) {
            if (requestCode == SEARCH_CODE) {

                if (data != null) {
                    boolean SEARCH_BY_CODE_byCode = data.getBooleanExtra(Constants.SEARCH_BY_CODE_byCode, false);
                    String CODE = data.getStringExtra(Constants.CODE);
                    FragmentManager fragmentManager = getParentFragmentManager();
                    SearchFragment searchFragment = new SearchFragment();
                    Bundle bundle = new Bundle();
                    bundle.putString(Constants.CODE, CODE);
                    bundle.putBoolean(Constants.SEARCH_BY_CODE_byCode, SEARCH_BY_CODE_byCode);
                    searchFragment.setArguments(bundle);
                    fragmentManager.beginTransaction().replace(R.id.mainContainer, searchFragment, "searchFragment").commitAllowingStateLoss();

                }


            }

        }
    }
}

File: app/src/main/java/com/ramez/shopp/fragments/CategoryProductsFragment.kt
package com.ramez.shopp.fragments


import android.Manifest
import android.app.Activity
import android.content.ContentValues
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.appbar.AppBarLayout
import com.karumi.dexter.Dexter
import com.karumi.dexter.PermissionToken
import com.karumi.dexter.listener.PermissionDeniedResponse
import com.karumi.dexter.listener.PermissionGrantedResponse
import com.karumi.dexter.listener.PermissionRequest
import com.karumi.dexter.listener.single.PermissionListener
import com.kcode.permissionslib.main.OnRequestPermissionsCallBack
import com.kcode.permissionslib.main.PermissionCompat
import com.ramez.shopp.ApiHandler.DataFeacher
import com.ramez.shopp.ApiHandler.DataFetcherCallBack
import com.ramez.shopp.classes.*
import com.ramez.shopp.Models.*
import com.ramez.shopp.R
import com.ramez.shopp.activities.FullScannerActivity
import com.ramez.shopp.activities.ProductDetailsActivity
import com.ramez.shopp.adapter.MainCategoryAdapter
import com.ramez.shopp.adapter.MainCategoryAdapter.OnMainCategoryItemClicked
import com.ramez.shopp.adapter.ProductCategoryAdapter
import com.ramez.shopp.adapter.SubCategoryAdapter
import com.ramez.shopp.databinding.FragmentCategoryProductsBinding
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.*


class CategoryProductsFragment : FragmentBase(), ProductCategoryAdapter.OnItemClick,
        OnMainCategoryItemClicked {
    var productList: ArrayList<ProductModel>? = null
    var data: ArrayList<AutoCompleteModel>? = null
    var autoCompleteList: ArrayList<String>? = null
    var mainCategoryDMS: ArrayList<CategoryModel>? = null
    var subCategoryDMS: ArrayList<ChildCat>? = null
    var gridLayoutManager: GridLayoutManager? = null
    var numColumn = 2
    var selectedSubCat = 0
    var categoryId = 0
    var countryId = 0
    var cityId = 0
    private var userId = "0"
    private var catPosition = 0
    private val filter = ""
    private var user: MemberModel? = null
    private var localModel: LocalModel? = null
    private var adapter: ProductCategoryAdapter? = null
    private val searchCode = 2000
    lateinit var binding: FragmentCategoryProductsBinding
    private var scanLauncher: ActivityResultLauncher<Intent>? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentCategoryProductsBinding.inflate(inflater, container, false)
        productList = ArrayList()
        mainCategoryDMS = ArrayList()
        gridLayoutManager = GridLayoutManager(activityy, numColumn)
        binding.productsRv.layoutManager = gridLayoutManager
        binding.listShopCategories.layoutManager =
            LinearLayoutManager(activityy, LinearLayoutManager.HORIZONTAL, false)
        binding.listSubCategory.layoutManager =
            LinearLayoutManager(activityy, LinearLayoutManager.HORIZONTAL, false)
        binding.productsRv.setHasFixedSize(true)
        binding.productsRv.itemAnimator = null
        data = ArrayList()
        autoCompleteList = ArrayList()
        localModel =
            if (UtilityApp.getLocalData() != null) UtilityApp.getLocalData() else UtilityApp.getDefaultLocalData(
                activityy
            )
        countryId = localModel?.countryId ?: Constants.default_country_id
        if (UtilityApp.isLogin() && UtilityApp.getUserData() != null && UtilityApp.getUserData().id != null) {
            user = UtilityApp.getUserData()
            userId = user?.id.toString()
        }
        cityId = localModel?.cityId?.toInt() ?: Constants.default_storeId.toInt()


        intentExtra
        initListeners()

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        scanLauncher = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) { result: ActivityResult? ->
            if (result?.resultCode == Activity.RESULT_OK) {
                val intentBundle = result.data?.extras
                val searchByCode = intentBundle?.getBoolean(Constants.SEARCH_BY_CODE_byCode, false)
                val code = intentBundle?.getString(Constants.CODE)

                if (!EventBus.getDefault().isRegistered(requireActivity()))
                    EventBus.getDefault().register(requireActivity())

                val bundle = Bundle()
                bundle.putString(Constants.CODE, code)
                bundle.putBoolean(Constants.SEARCH_BY_CODE_byCode, searchByCode!!)
                bundle.putInt(
                    Constants.KEY_FRAGMENT_ID,
                    R.id.searchFragment
                )
                EventBus.getDefault()
                    .post(MessageEvent(MessageEvent.TYPE_FRAGMENT, bundle))

//                val searchFragment = SearchFragment()
//                searchFragment.arguments = bundle
//                fragmentManager.beginTransaction()
//                    .replace(R.id.mainContainer, searchFragment, "searchFragment")
//                    .commitAllowingStateLoss()
            }

        }


    }


    private fun initListeners() {
        binding.headerAppBarLY.addOnOffsetChangedListener(object : AppBarStateChangeListener() {
            override fun onStateChanged(appBarLayout: AppBarLayout, state: State) {
                if (state == State.COLLAPSED) {
                    EventBus.getDefault().post(MessageEvent(MessageEvent.TYPE_SHOW_SEARCH, true))
                } else {
                    EventBus.getDefault().post(MessageEvent(MessageEvent.TYPE_SHOW_SEARCH, false))
                }
            }

            override fun onBarOffsetChanged(appBarLayout: AppBarLayout, offset: Int) {}
        })
        binding.failGetDataLY.refreshBtn.setOnClickListener {
            getProductList(
                selectedSubCat,
                countryId,
                cityId,
                userId,
                filter,
                0,
                10
            )
        }
        binding.searchBut.setOnClickListener {
            val bundle = Bundle()
            bundle.putInt(Constants.KEY_FRAGMENT_ID, R.id.searchFragment)
            EventBus.getDefault()
                .post(MessageEvent(MessageEvent.TYPE_FRAGMENT, bundle))
        }
        binding.barcodeBut.setOnClickListener { checkCameraPermission() }
    }

    fun initAdapter() {
        adapter = ProductCategoryAdapter(
            activityy,
            binding.productsRv,
            productList,
            selectedSubCat,
            countryId,
            cityId,
            userId,
            0,
            "",
            this,
            numColumn,
            0
        )
        binding.productsRv.adapter = adapter
    }

    override fun onItemClicked(position: Int, productModel: ProductModel) {
        val intent = Intent(activityy, ProductDetailsActivity::class.java)
        intent.putExtra(Constants.DB_productModel, productModel)
        startActivity(intent)
    }

    private fun cancelAPiCall() {
        if (adapter != null && adapter!!.apiCall != null && adapter!!.apiCall.isExecuted) {
            adapter!!.isCanceled = true
            adapter!!.apiCall.cancel()
        }
    }

    private val intentExtra: Unit
        get() {
            val bundle = arguments
            Log.i(javaClass.simpleName, "Log getIntentExtra ")
            if (bundle != null) {
                mainCategoryDMS = bundle.getSerializable(Constants.CAT_LIST) as ArrayList<CategoryModel>?
                //            boolean subCat = bundle.getBoolean(Constants.SUB_CAT_ID);

                catPosition = bundle.getInt(Constants.position, 0)

                if (mainCategoryDMS == null) {
                    if (UtilityApp.getCategories() != null && UtilityApp.getCategories().size > 0) {
                        mainCategoryDMS = UtilityApp.getCategories()

                    } else {
                        getCategories(localModel!!.cityId.toInt())
                    }
                }

                Log.i(javaClass.simpleName, "Log catPosition $catPosition")

                categoryId = bundle.getInt(Constants.MAIN_CAT_ID, 0)
                selectedSubCat = bundle.getInt(Constants.SUB_CAT_ID, 0)

                Log.i(javaClass.simpleName, "Log mainCategoryDMS " + mainCategoryDMS!!.size)
                Log.i(javaClass.simpleName, "Log category_id $categoryId")
                Log.i(javaClass.simpleName, "Log sub_cat_id $selectedSubCat")
                if (categoryId == 0 && selectedSubCat != 0) {
                    getCatIdFromSub(selectedSubCat)
                } else {
                    initData()
                }
            }
        }

    private fun initData() {
        initMainCategoryAdapter()
        initSubCatList()
        getProductList(selectedSubCat, countryId, cityId, userId, filter, 0, 10)
    }

    private fun initSubCatList() {
        val childCat = ChildCat()
        childCat.id = categoryId
        childCat.hName = getString(R.string.all)
        childCat.name = getString(R.string.all)
        if (selectedSubCat == 0) selectedSubCat = categoryId
        if (subCategoryDMS == null) searchSubCatList()
        subCategoryDMS!!.add(0, childCat)
        initSubCategoryAdapter()
    }

    private fun initSubCategoryAdapter() {
        val subCategoryAdapter = SubCategoryAdapter(
            activityy, subCategoryDMS, selectedSubCat
        ) { `object`: ChildCat ->
            selectedSubCat = `object`.id
            cancelAPiCall()
            getProductList(selectedSubCat, countryId, cityId, userId, "", 0, 10)
        }
        binding.listSubCategory.adapter = subCategoryAdapter
    }

    private fun initMainCategoryAdapter() {
        val mainCategoryShopAdapter = MainCategoryAdapter(activityy, mainCategoryDMS, this, categoryId)
        binding.listShopCategories.adapter = mainCategoryShopAdapter
        getCatPosition()

    }

    private fun getProductList(
        category_id: Int,
        country_id: Int,
        city_id: Int,
        user_id: String?,
        filter: String?,
        page_number: Int,
        page_size: Int
    ) {
        binding.loadingProgressLY.loadingProgressLY.visibility = View.VISIBLE
        binding.productsRv.visibility = View.GONE
        binding.noDataLY.noDataLY.visibility = View.GONE
        binding.failGetDataLY.failGetDataLY.visibility = View.GONE
        DataFeacher(false,
            object : DataFetcherCallBack {
                override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                    if (isVisible) {
                        val result = obj as FavouriteResultModel?
                        var message: String? = getString(R.string.fail_to_get_data)
                        binding.loadingProgressLY.loadingProgressLY.visibility = View.GONE
                        if (func == Constants.ERROR) {
                            if (result != null && result.message != null) {
                                message = result.message
                            }
                            binding.productsRv.visibility = View.GONE
                            binding.noDataLY.noDataLY.visibility = View.GONE
                            binding.failGetDataLY.failGetDataLY.visibility = View.VISIBLE
                            binding.failGetDataLY.failTxt.text = message
                        } else if (func == Constants.NO_CONNECTION) {
                            binding.failGetDataLY.failGetDataLY.visibility = View.VISIBLE
                            binding.failGetDataLY.failTxt.setText(R.string.no_internet_connection)
                            binding.failGetDataLY.noInternetIv.visibility = View.VISIBLE
                            binding.productsRv.visibility = View.GONE
                        } else {
                            if (IsSuccess) {
                                if (result!!.data != null && result.data.size > 0) {
                                    binding.productsRv.visibility = View.VISIBLE
                                    binding.noDataLY.noDataLY.visibility = View.GONE
                                    binding.failGetDataLY.failGetDataLY.visibility = View.GONE
                                    productList = ArrayList(result.data)
                                    initAdapter()
                                } else {
                                    binding.productsRv.visibility = View.GONE
                                    binding.noDataLY.noDataLY.visibility = View.VISIBLE
                                }
                            } else {
                                binding.productsRv.visibility = View.GONE
                                binding.noDataLY.noDataLY.visibility = View.GONE
                                binding.failGetDataLY.failGetDataLY.visibility = View.VISIBLE
                                binding.failGetDataLY.failTxt.text = message
                            }
                        }
                    }
                }

            }).getFavorite(category_id, country_id, city_id, user_id, filter, 0, page_number, page_size)
    }

    override fun OnMainCategoryItemClicked(mainCategoryDM: CategoryModel, position: Int) {
        categoryId = mainCategoryDM.id
        selectedSubCat = categoryId
        subCategoryDMS = ArrayList(mainCategoryDM.childCat)
        initSubCatList()
        cancelAPiCall()
        getProductList(selectedSubCat, countryId, cityId, userId, filter, 0, 10)
    }

    private fun checkCameraPermission() {
        Dexter.withContext(activityy).withPermission(Manifest.permission.CAMERA)
            .withListener(object : PermissionListener {
                @RequiresApi(api = Build.VERSION_CODES.M)
                override fun onPermissionGranted(response: PermissionGrantedResponse) {
                    startScan()
                }

                override fun onPermissionDenied(response: PermissionDeniedResponse) {
                    Toast.makeText(
                        activityy,
                        "" + getString(R.string.permission_camera_rationale),
                        Toast.LENGTH_SHORT
                    ).show()
                }

                override fun onPermissionRationaleShouldBeShown(
                    permission: PermissionRequest,
                    token: PermissionToken
                ) {
                    token.continuePermissionRequest()
                }
            }).withErrorListener {
                Toast.makeText(
                    activityy,
                    "" + getString(R.string.error_in_data),
                    Toast.LENGTH_SHORT
                ).show()
            }.onSameThread().check()
    }



    private fun startScan() {
        try {
            val builder = PermissionCompat.Builder(activityy)
            builder.addPermissions(arrayOf(Manifest.permission.CAMERA))
            builder.addPermissionRationale(getString(R.string.should_allow_permission))
            builder.addRequestPermissionsCallBack(object : OnRequestPermissionsCallBack {
                override fun onGrant() {
                    val intent = Intent(activityy, FullScannerActivity::class.java)
                    scanLauncher?.launch(intent)

                }

                override fun onDenied(permission: String) {
                    Toast(R.string.some_permission_denied)
                }
            })
            builder.build().request()
        } catch (var2: Exception) {
            var2.printStackTrace()
        }
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: MessageEvent) {
        if (event.type == MessageEvent.TYPE_view) {
            numColumn = event.data as Int
            initAdapter()
            gridLayoutManager!!.spanCount = numColumn
            adapter!!.notifyDataSetChanged()
        }
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        EventBus.getDefault().unregister(this)
    }


    fun getCategories(storeId: Int) {
        DataFeacher(false, object : DataFetcherCallBack {
            override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                val result = obj as CategoryResultModel
                if (IsSuccess) {
                    if (result.data != null && result.data.size > 0) {
                        mainCategoryDMS = result.data
                        Log.i(ContentValues.TAG, "Log productBestList" + mainCategoryDMS?.size)
                        UtilityApp.setCategoriesData(mainCategoryDMS)
                    }
                }
            }

        }).GetAllCategories(storeId)
    }

    private fun getCatIdFromSub(subId: Int) {
        for (i in mainCategoryDMS!!.indices) {
            if (subId == mainCategoryDMS!![i].id) {
                categoryId = mainCategoryDMS!![i].id
                selectedSubCat = categoryId // this for all cat
                subCategoryDMS = ArrayList(mainCategoryDMS!![i].childCat)
                initData()
                break
            }
            val subArrayList = mainCategoryDMS!![i].childCat
            for (j in subArrayList.indices) {
                if (subId == subArrayList[j].id) {
                    categoryId = mainCategoryDMS!![i].id
                    Log.i(javaClass.simpleName, "Log subCat category_id $categoryId")
                    selectedSubCat = subId
                    subCategoryDMS = ArrayList(mainCategoryDMS!![i].childCat)
                    initData()
                    break
                }
            }
        }
    }

    private fun searchSubCatList() {

        if (mainCategoryDMS != null)
            for (i in mainCategoryDMS!!.indices) {
                if (categoryId == mainCategoryDMS!![i].id) {
                    subCategoryDMS = ArrayList(mainCategoryDMS!![i].childCat)
                    break
                }
            }
        if (subCategoryDMS == null) subCategoryDMS = ArrayList()
    }

    companion object {
        private const val ZBAR_CAMERA_PERMISSION = 1
    }

    private fun getCatPosition() {

        Log.i(javaClass.simpleName, "Log catPosition categoryId$categoryId")

        for ((i, categoryModel) in mainCategoryDMS?.withIndex() ?: mutableListOf()) {
            if (categoryId == categoryModel.id) {
                catPosition = i
                Log.i(javaClass.simpleName, "Log catPosition getCatPosition$catPosition")
                break
            }
        }

        binding.listShopCategories.scrollToPosition(catPosition)


    }


}